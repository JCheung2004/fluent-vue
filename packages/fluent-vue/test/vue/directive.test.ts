import { createLocalVue, mount } from '@vue/test-utils'

import { FluentBundle, FluentResource } from '@fluent/bundle'
import ftl from '@fluent/dedent'

import FluentVue from '../../src'

describe('directive', () => {
  let options: any
  let bundle: FluentBundle

  beforeEach(() => {
    const localVue = createLocalVue()
    localVue.use(FluentVue)

    bundle = new FluentBundle('en-US')

    const fluent = new FluentVue({
      bundles: [bundle]
    })

    options = {
      fluent,
      localVue
    }
  })

  it('translates text content', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Link text
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link href="/foo">Fallback text</a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('warns about missing key arg', () => {
    // Arrange
    const component = {
      template: `<a v-t href="/foo">Fallback text</a>`
    }

    const warn = jest.fn()
    console.warn = warn

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
    expect(warn).toHaveBeenCalledTimes(1)
  })

  it('can use parameters', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Hello {$name}
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link="{ name }" href="/foo">Fallback text</a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('can translate DOM attributes', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Hello {$name}
        .aria-label = Localized aria
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link.aria-label="{ name }" href="/foo" aria-label="Fallback aria">Fallback text</a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('automatically binds whitelisted attrs', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Text
        .aria-label = Hello {$name}
        .not-allowed = Not allowed attrs value
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link="{ name }">Fallback</a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('works without fallbacks', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Hello {$name}
        .aria-label = Localized aria
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link.aria-label="{ name }"></a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('updates translations on component update', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Hello {$name}
        .aria-label = Localized aria
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link.aria-label="{ name }"></a>`
    }

    const mounted = mount(component, options)

    expect(mounted).toMatchSnapshot()

    // Act
    mounted.setData({ name: 'Anna' })

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('preserves translations on component update', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link =
        .aria-label = Hello {$name}
      `)
    )

    const component = {
      data: () => ({
        name: 'John',
        otherName: 'Anna'
      }),
      template: `<a v-t:link.aria-label="{ name }">{{ otherName }}</a>`
    }

    const mounted = mount(component, options)

    expect(mounted).toMatchSnapshot()

    // Act
    mounted.setData({ otherName: 'Test' })

    // Assert
    expect(mounted).toMatchSnapshot()
  })

  it('works with multiple attributes', () => {
    // Arrange
    bundle.addResource(
      new FluentResource(ftl`
      link = Text
        .aria-label = Hello {$name}
        .placeholder = Placeholder
      `)
    )

    const component = {
      data: () => ({
        name: 'John'
      }),
      template: `<a v-t:link.aria-label.placeholder="{ name }">Fallback</a>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()
  })
})
