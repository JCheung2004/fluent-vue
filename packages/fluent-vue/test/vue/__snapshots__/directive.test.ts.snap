// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directive automatically binds whitelisted attrs 1`] = `<a aria-label="Hello ⁨John⁩">Text</a>`;

exports[`directive can translate DOM attributes 1`] = `<a href="/foo" aria-label="Localized aria">Hello ⁨John⁩</a>`;

exports[`directive can use parameters 1`] = `<a href="/foo">Hello ⁨John⁩</a>`;

exports[`directive preserves translations on component update 1`] = `<a aria-label="Hello ⁨John⁩">Anna</a>`;

exports[`directive preserves translations on component update 2`] = `<a aria-label="Hello ⁨John⁩">Test</a>`;

exports[`directive translates text content 1`] = `<a href="/foo">Link text</a>`;

exports[`directive updates translations on component update 1`] = `<a aria-label="Localized aria">Hello ⁨John⁩</a>`;

exports[`directive updates translations on component update 2`] = `<a aria-label="Localized aria">Hello ⁨Anna⁩</a>`;

exports[`directive warns about missing key arg 1`] = `<a href="/foo">Fallback text</a>`;

exports[`directive works with multiple attributes 1`] = `<a aria-label="Hello ⁨John⁩" placeholder="Placeholder">Text</a>`;

exports[`directive works without fallbacks 1`] = `<a aria-label="Localized aria">Hello ⁨John⁩</a>`;
